tests/code_size/hello_webgl2_fastcomp_asmjs.json

```diff
@@ -1,12 +1,12 @@
 {
   "a.html": 580,
   "a.html.gz": 382,
-  "a.js": 5322,
-  "a.js.gz": 2505,
+  "a.js": 5565,
```

Comment:
an almost 5% code size increase here is significant. what is that because of specifically?

Reply:
(If we can't avoid a code size increase here, we'll need to get consensus that it is worth the regression, which will take longer.)

Reply:
It appears that about half of that may be due to the new variable names not getting shortened... 😐 Any ideas on how to create/declare those without needing to make externs and bypassing the shortening optimization?

Reply:
I tried moving them to the shell files, following how `Module` is declared. It reduced the size a bit, but not as much as I thought I was seeing in my manual testing. Please let me know if this is still an issue.

Reply:
K, just pushed another change moving them back in order for the tests to pass. I'm sorry for all my chatter while I figure this out.

The size increase is indeed coming from the longer names. My attempt to put the variable declarations into shell.js and shell_minimal.js didn't work because the resolve and reject function variables only get renamed where they're used, not where they're created -- the code created as part of the `modularize` function. And since that code doesn't get optimized, there's still some extra bytes there, which explains the smaller-than-expected size decrease.

It seems to me that it would be ideal if we could run Closure Compiler after the generated code has been modularized. I tried naively doing that, moving the call to `modularize` earlier in `run`, but that didn't work because some variables were renamed that we don't want to be, namely `Module`.

An alternative, less ideal solution to shave off bytes would be to use really short variable names in the source. We would run the risk of having a naming collision with any short variable names that Closure Compiler comes up with, and of course the sacrifice to readability.

Reply:
I don't remember offhand why we run closure in the order we do, but it's probably not easy to change, that's a workflow that's been carefully tuned.

I don't understand the issue here yet, though. Can you perhaps show me examples of what you mean here? Like code samples (with proper whitespace, so `-g1`, but otherwise with closure) before and after?

Reply:
Yes, I'd be happy to. I've been struggling to figure out how exactly to do that though... I've been relying on hacky changes in runner.py to save the generated files because using `EMTEST_SAVE_DIR` didn't actually do anything for me. Could you please give me an example command to run, or point me to where in the docs I can read to do it myself?

Reply:
Very odd `EMTEST_SAVE_DIR=1` doesn't work for you. Do you not get similiar results when running these commands?

```bash
$ EMTEST_SAVE_DIR=1 ./tests/runner.py  test_loop
Running test_core: (1 tests)
test_loop (test_core.wasm0) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.614s

OK
$ ls -al /tmp/emscripten_test/
total 796
drwxr-xr-x  2 azakai primarygroup   4096 Apr 23 17:00 .
drwxrwxrwt 55 root   root         655360 Apr 23 17:00 ..
-rw-r--r--  1 azakai primarygroup    490 Apr 23 17:00 src.c
-rw-r--r--  1 azakai primarygroup 115013 Apr 23 17:00 src.c.o.js
-rw-r--r--  1 azakai primarygroup  21974 Apr 23 17:00 src.c.o.wasm
-rw-r--r--  1 azakai primarygroup      0 Apr 23 17:00 stderr
-rw-r--r--  1 azakai primarygroup      7 Apr 23 17:00 stdout
```

However, I think for a test like minimal_runtime_code_size that creates more than one file, it will overwrite the first. It may be simplest to either make the test halt after the point you want (add a python `assert 0` after the last line you want, say), or instead you can just run a similar command manually, separately from the test suite. I imagine building with `-s MINIMAL_RUNTIME -s MODULARIZE --closure 1 -Os` with and without this PR might show the relevant differences.

Reply:
Ok, I'll try that out, thank you.

Yeah for some reason using `EMTEST_SAVE_DIR` as you showed puts my files in `/var/folders/7y/pk5bt2qn2s98q3w1tbf5q89w00j4ht/T/emscripten_test/`, which isn't really any different than if I don't use the option. It seems a random directory like that is the default for macOS? Based on some cursory googling ([see this article](https://www.cleverfiles.com/help/temp-folders-mac.html)), the OS sets the `TMPDIR` environment variable to that weird path, which is what the `tempfile` Python module is pulling from

Reply:
Here's the diff of the generated `a.js` file of the `hello_webgl_sources` case of `test_minimal_runtime_code_size` -- https://www.diffchecker.com/FgaKg6Su.

The only difference is in the promise setup and the return. There's no difference in any of the other files. This shows the size increase is coming down to the variable names.

Reply:
I think if we separate the first part of the `modularize` function that wraps the src in a factory function from the latter part that handles generating the export logic, we could do the wrapping before running optimizations. There's even the existing function `module_export_name_substitution` called right after `modularize` that we could broaden to something like `generate_module_exports` and combine the export generating logic there.

Reply:
If that's possible then it sounds like it could help, yeah. However, it also sounds like it would reduce the issue (minify the name) but we would still have a bunch of new code here, and regress code size.

Given the almost 5% code size regression here, I think we would want to put this behind a flag, if we can't get rid of almost all of it.

Reply:
I tried moving the modularization before the optimizer and ran into this very specific and very helpful error

```
could not find the assigment to "asmLibraryArg". perhaps --pre-js or --post-js code moved it out of the global scope? (things like that should be done after emcc runs, as they do not need to be run through the optimizer which is the special thing about --pre-js/--post-js code)
```

So it won't be as simple as I'd hoped 😕

Reply:
Yeah, the minifier does depend on not having extra scopes when it runs, as it has some assumptions.

We may be able to run the minifier an additional time again, after everything, if it matters. But we can leave that for later.